osquery/tables/system/darwin/signature.mm

```diff
@@ -0,0 +1,126 @@
+                         QueryData& results) {
+  Row r;
+  OSStatus result = 0;
+  CFStringRef ident = NULL;
+  CFDictionaryRef codeInfo = NULL;
+  SecStaticCodeRef staticCode = NULL;
+  SecCSFlags csFlags = kSecCSDefaultFlags;
+
+  // Defaults
+  r["path"] = path;
+  r["signed"] = INTEGER(0);
+  r["identifier"] = "";
+
+  // Create a URL that points to this file.
+  auto url = (__bridge CFURLRef)[NSURL fileURLWithPath:@(path.c_str())];
+  if (url == nullptr) {
+    VLOG(1) << "Could not create URL from file: " << path;
+    goto finally;
+  }
+
+  // Create the static code object.
+  result = SecStaticCodeCreateWithPath(url, kSecCSDefaultFlags, &staticCode);
+  if (result != 0) {
+    VLOG(1) << "Could not create static code object for file: " << path;
+    goto finally;
+  }
+
+  // Set up the flags - some of them aren't present on 10.8.
+  if (floor(NSAppKitVersionNumber) > NSAppKitVersionNumber10_8) {
+    csFlags = kSecCSDefaultFlags | kSecCSCheckNestedCode;
+  }
+  else {
+    csFlags = kSecCSBasicValidateOnly;
+  }
+
+  // Actually validate.
+  result = SecStaticCodeCheckValidityWithErrors(staticCode, csFlags, NULL, NULL);
+  if (result != 0) {
+    // If this errors, then we either don't have a signature, or it's malformed.
+    VLOG(2) << "Static code validity check failed for file: " << path;
+    goto done;
+  }
+
+  result = SecCodeCopySigningInformation(
+    staticCode,
+    kSecCSSigningInformation | kSecCSRequirementInformation,
+    &codeInfo);
+  if (result != 0) {
+    VLOG(1) << "Could not get signing information for file: " << path;
+    goto finally;
+  }
+
+  // If we don't get an identifier for this file, then it's not signed.
+  ident = (CFStringRef)CFDictionaryGetValue(codeInfo, kSecCodeInfoIdentifier);
+  if (ident == nullptr) {
+    VLOG(2) << "No identifier found for file: " << path;
+    goto done;
+  }
+
+  // We have an identifier - this indicates that the file is signed, and, since
+  // it didn't error above, it's *also* a valid signature.
+  r["signed"] = INTEGER(1);
+  r["identifier"] = stringFromCFString(ident);
+
+  // TODO(andrew-d): can get more information from the signature here?
+
+  VLOG(3) << "Everything succeeded for: " << path;
+
+done:
+  results.push_back(r);
+
+finally:
+  if (staticCode) CFRelease(staticCode);
+  if (codeInfo) CFRelease(codeInfo);
+}
+
+QueryData genSignature(QueryContext& context) {
+  QueryData results;
+
+  // The query must provide a predicate with constraints including path or
+  // directory. We search for the parsed predicate constraints with the equals
+  // operator.
+  auto paths = context.constraints["path"].getAll(EQUALS);
+  for (const auto& path_string : paths) {
+    boost::filesystem::path path = path_string;
+
+    // Note: we are explicitly *not* using is_regular_file here, since you can
+    // pass a directory path to the verification functions (e.g. for app
+    // bundles, etc.)
+    if (!boost::filesystem::exists(path)) {
```

Comment:
Instead of using boost's `::exists` directly, prefer passing the `const std::string` to [`osquery::pathExists`](https://github.com/facebook/osquery/blob/master/osquery/filesystem/filesystem.cpp#L178) as you'll get exception handling for free.